Index: process.c
===================================================================
--- process.c.orig
+++ process.c
@@ -1296,8 +1296,11 @@ remove_entry(char *inputfilename, int li
     /*
      * unlink the auth we were asked to
      */
-    while (!eq_auth((list = *listp)->auth, auth))
+    while (!eq_auth((list = *listp)->auth, auth)) {
 	listp = &list->next;
+        if (!*listp)
+            return 0;
+    }
     *listp = list->next;
     XauDisposeAuth (list->auth);                    /* free the auth */
     free (list);				    /* free the link */
